### ISSUE TITLE
Incorrect NaN gradient from distribution.Normal.log_prob when using subset

### ISSUE THREAD

**Issue by samedii** [pytorch/pytorch#15131]
## 🐛 Bug
When a subset of a log_prob tensor is NaN then you can select the subset that is not NaN. This should then result in a finite gradient (in many cases) when doing backprop through the distribution. It does not

## To Reproduce

```
import numpy as np
import torch
import torch.nn as nn
import torch.distributions as dist
x = torch.tensor([1.0,2,3,np.nan])
y = torch.tensor([1.0,2,3,4])

k = nn.Parameter(0.01*torch.randn(1))

d = dist.Normal(loc=k*x, scale=1)

log_prob = d.log_prob(y)
print(log_prob)
# tensor([-1.4253, -2.9444, -5.4763,     nan], grad_fn=<SubBackward>)

loss = -log_prob[:-1].mean()
print(loss)
# tensor(3.2820, grad_fn=<NegBackward>)

loss.backward()
print(k.grad)
# tensor([nan])
```

## Expected behavior
Should see a finite gradient.

## Environment

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0

OS: Microsoft Windows 7 Enterprise
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy (1.13.3)
[pip] numpydoc (0.7.0)
[conda] mkl                       2018.0.0             h36b65af_4
[conda] mkl-service               1.1.2            py36h57e144c_4

**Comment by soumith** [comment]
cc: distributions folks @vishwakftw @neerajprad @fritzo

**Comment by vishwakftw** [comment]
I think this is an autograd issue and not specific to distributions. The following code example supports my hypothesis:

```python
import torch
a = torch.randn(4)
a[-1] = float('nan')
a.requires_grad_()  # tensor([-1.6708,  0.6398, -0.9565,     nan], requires_grad=True)
b = torch.randn(4)  # tensor([ 0.0499, -0.1655,  0.9131,  1.4798])
c = (a - b)**2  # tensor([2.9609, 0.6485, 3.4954,    nan], grad_fn=<PowBackward0>)
l = c[:-1].sum()
l.backward()
a.grad  # tensor([-3.4414,  1.6106, -3.7392,     nan])
```

If I understand correctly, I think the last item of `a.grad` should be 0.

**Comment by neerajprad** [comment]
Expanding on what @vishwakftw mentioned, the failure case below is similar. Surprisingly, the first case without exponentiation works okay, but the `nan` ends up getting leaked after `.pow`.

```python

>>> import torch
>>> a = torch.ones(4)
>>> a[-1] = float('NaN')
>>> print(a)
tensor([1., 1., 1., nan])

>>> b = torch.tensor(1., requires_grad=True)
# This works as expected - nan is masked
>>> (a + b)[:-1].sum().backward()   
>>> print(b.grad)   
tensor(3.)

>>> b.grad.zero_()
# nan is leaked with exponentiation.
>>> (a + b).pow(2)[:-1].sum().backward()
>>> print(b.grad)   
tensor(nan)
```

**Comment by t-vi** [comment]
What happens is that the indexing will have a tensor matching the full tensor as grad_in. As expected the last (i.e. the unused) element grad_in will have `0` gradients.
Now, any operation that uses the `NaN` input to compute its grad_in  from grad_out (like mul, or pow) will see something `0*NaN` and that is most reasonably evaluated as `NaN` in general. If you had a way of saying 

Suprising? Perhaps.
Wrong? Not as much as one might think at first sight.
Now if you had a way of saying "this is strongly-0 that trumps NaN", it would be great, but we don't.

The common solution of doing masking more strongly probably is good here, too: For this you would zero out gradients in a backward hook on k*x.

**Comment by samedii** [comment]
Using the backward hook is not quite so easy though as you need to create a layer in between to stop the gradient. It seems easier to just set the NaN value in the data to 0 and then do the masking. Everything should be ok then unless I am missing something.

Backward hook with "layer" to catch NaNs in case someone needs it:
```
import numpy as np
import torch
import torch.nn as nn

def create_hook(name):
    def hook(grad):
        print(name, grad)
    return hook

x = torch.tensor([1.0, np.nan])
k = nn.Parameter(0.01*torch.randn(1))

between = k.repeat(2) # need the extra dimensions in order to fix the NaN gradient
between.register_hook(create_hook('between'))

y = between*x

masked = y[:-1]

loss = masked.sum()
print(f'loss: {loss}')

loss.backward()
```

Closing this as it seems it would be difficult to find a less hacky solution
